cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/executor/ImpersonatingDatasetAdmin.java

```diff
@@ -0,0 +1,133 @@
+import java.util.concurrent.Callable;
+
+/**
+ * A {link DatasetAdmin} that executes operations, while impersonating.
+ */
+class ImpersonatingDatasetAdmin implements DatasetAdmin {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ImpersonatingDatasetAdmin.class);
+
+  private final DatasetAdmin delegate;
+  private final Impersonator impersonator;
+  private final NamespaceId namespaceId;
+
+  ImpersonatingDatasetAdmin(DatasetAdmin delegate, Impersonator impersonator, NamespaceId namespaceId) {
+    this.delegate = delegate;
+    this.impersonator = impersonator;
+    this.namespaceId = namespaceId;
+  }
+
+  @Override
+  public boolean exists() throws IOException {
+    return execute(new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        return delegate.exists();
+      }
+    });
+  }
+
+  @Override
+  public void create() throws IOException {
+    execute(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        delegate.create();
+        return null;
+      }
+    });
+  }
+
+  @Override
+  public void drop() throws IOException {
+    execute(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        delegate.drop();
+        return null;
+      }
+    });
+  }
+
+  @Override
+  public void truncate() throws IOException {
+    execute(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        delegate.truncate();
+        return null;
+      }
+    });
+  }
+
+  @Override
+  public void upgrade() throws IOException {
+    execute(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        delegate.upgrade();
+        return null;
+      }
+    });
+  }
+
+  @Override
+  public void close() throws IOException {
+    execute(new Callable<Void>() {
+      @Override
+      public Void call() throws Exception {
+        delegate.close();
+        return null;
+      }
+    });
+  }
+
+  // helper method to execute a callable, while declaring only IOException as being thrown
+  private <T> T execute(final Callable<T> callable) throws IOException {
+    try {
+      return impersonator.doAs(namespaceId, callable);
+    } catch (IOException ioe) {
+      throw ioe;
+    } catch (Throwable t) {
+      Throwables.propagateIfPossible(t);
+
+      // since the callables we execute only throw IOException (besides unchecked exceptions),
+      // this should never happen
+      LOG.warn("Unexpected exception while executing dataset admin operation in namespace {}.", namespaceId,  t);
+      // the only checked exception that the Callables in this class is IOException, and we handle that in the previous
+      // catch statement. So, no checked exceptions should be wrapped by the following statement. However, we need it
+      // because ImpersonationUtils#doAs declares 'throws Throwable', because it can throw other checked exceptions
```

Comment:
does it have to be ``throws Throwable``? Can it be ``throws Exception`` instead?